src/quilla/plugins/base_storage.py

```diff
@@ -0,0 +1,188 @@
+'''
+This module contains a base class that can be useful when defining new
+storage plugins, as there are some behaviours that will be shared among
+the configuration objects
+'''
+
+from abc import (
+    ABC,
+    abstractproperty,
+    abstractmethod
+)
+from typing import (
+    Optional
+)
+
+from quilla.common.enums import VisualParityImageType
+
+
+class BaseStorage(ABC):
+    @abstractproperty
+    def is_enabled(self) -> bool:
+        '''
+        A property to be used to determine if the plugin is configured and
+        therefore able to run
+
+        Returns:
+            True if the plugin is configured, False otherwise
+        '''
+
+    @abstractmethod
+    def find_image_by_baseline(self, baseline_id: str) -> bytes:
+        '''
+        Searches the defined storage method for the image matching some
+        baseline ID
+
+        Args:
+            baseline_id: The unique ID to search for. It is assumed every
+                baseline ID is unique regardless of what test is requesting it.
+
+        Returns:
+            The bytes representation of the stored image if found, or an empty
+            bytes object if the image is not found.
+        '''
+
+    @abstractmethod
+    def cleanup_reports(self):
+        '''
+        Searches for reports that match some cleanup criteria
+        '''
+
+    @abstractmethod
+    def store_treatment_image(
+        self,
+        run_id: str,
+        baseline_id: str,
+        treatment: bytes,
+    ) -> str:
+        '''
+        Stores a treatment image within the storage mechanism enabled by the plugin
+
+        Args:
+            run_id: The run ID of the current Quilla run, to version the treatment images
+            baseline_id: The ID of the baseline that this treatment image is associated
+                with
+            treatment: The image data in bytes
+
+        Returns:
+            An identifier that can locate the newly stored treatment image
+        '''
+
+    @abstractmethod
+    def store_baseline_image(
+        self,
+        baseline_id: str,
+        baseline: bytes,
+    ) -> str:
+        '''
+        Stores a baseline image under the given baseline_id.
+
+        This function should be used to update the current baseline
+        image, or create a new one if the baseline did not previously exist
+
+        Args:
+            baseline_id: A unique identifier for the image
+            baseline: The image data in bytes
+
+        Returns:
+            A URI for the new baseline image
+        '''
+
+    @abstractmethod
+    def make_baseline_uri(
+        self,
+        run_id: str,
```

Comment:
if find_image_by_baseline() and store_baseline_image() do not need run_id why make_baseline_uri() would need run_id? they need to be consistent.

Reply:
make_baseline_uri gives the opportunity to create a copy of the baseline first before returning the copy’s URI, which is why it includes the run_id- i can add the run ID to the other functions since that might assist with the versioning of baselines